hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/acl/OMVolumeAddAclRequest.java

```diff
@@ -0,0 +1,126 @@
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.om.request.volume.acl;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.ozone.OzoneAcl;
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+import org.apache.hadoop.ozone.om.OMMetrics;
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.helpers.OmVolumeArgs;
+import org.apache.hadoop.ozone.om.request.OMClientRequest;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.om.response.volume.OMVolumeAddAclResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.AddAclRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.AddAclResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Status;
+
+import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+import org.apache.hadoop.ozone.security.acl.OzoneObjInfo;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.VOLUME_LOCK;
+
+/**
+ * Handles add acl request.
+ */
+public class OMVolumeAddAclRequest extends OMClientRequest {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMVolumeAddAclRequest.class);
+
+  public OMVolumeAddAclRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex) {
+    AddAclRequest addAclRequest = getOmRequest().getAddAclRequest();
+    Preconditions.checkNotNull(addAclRequest);
+
+    OMResponse.Builder omResponse = OMResponse.newBuilder().setCmdType(
+        OzoneManagerProtocolProtos.Type.AddAcl).setStatus(
+            Status.OK).setSuccess(true);
+
+    if (!addAclRequest.hasAcl()) {
+      omResponse.setStatus(Status.INVALID_REQUEST).setSuccess(false);
+      return new OMVolumeAddAclResponse(null, omResponse.build());
+    }
+
+    OzoneObjInfo ozoneObj = OzoneObjInfo.fromProtobuf(addAclRequest.getObj());
```

Comment:
If OmVolumeArgs has this direct protobuf structure's we can save the protobuf conversion. Previously this might be required because we used to have individual methods addACL with internal OzoneAcl structure of OM. As now we have implementation logic for each request, we can save this protobuf conversion.

Reply:
Can you elaborate on the OmVolumeArgs protobuf conversion? OzoneManager does not take VolumeArgs directly. All volume/bucket/key are wrapped into OzoneObj. Even though we have separate Requests, but OzoneObj from RPC layer has to be converted to specific structure.

Reply:
ozoneObj is used to getVolumeName , we can directly getVolumeName from addAclRequest.getObj().getPath(), because we already checked if resType is Volume, then only we shall be in this place.

In similar way, we can avoid conversion for  OzoneAcl ozoneAcl = OzoneAcl.fromProtobuf(addAclRequest.getAcl());
If OmVolumeArgs can directly store List<OzoneAclInfo>. In this way we can avoid protobuf conversion.

Reply:
I agree that the first protoc conversion can be avoided. 
But the second change from OzoneAclMap to OzoneAclInfo in OmVolumeArgs has many references. It is preferred to do this optimization in a separate JIRA.